src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py

```diff
@@ -1039,6 +1039,45 @@ def create_or_replace_dynamic_table(
             source_plan,
         )
 
+    def _merge_file_format_options(
+        self, file_format_options: Dict[str, Any], options: Dict[str, str]
+    ) -> Dict[str, Any]:
+        """
+        Merges remotely defined file_format options with the local options set. This allows the client
+        to override any locally defined options that are incompatible with the file format.
+        """
+        if "FORMAT_NAME" not in options:
+            return file_format_options
+
+        def process_list(list_property):
+            split_list = list_property.lstrip("[").rstrip("]").split(", ")
+            if len(split_list) == 1:
+                return split_list[0]
+            return tuple(split_list)
+
+        type_map = {
+            "String": str,
+            "List": process_list,
+            "Integer": int,
+            "Boolean": bool,
+        }
+        new_options = {**file_format_options}
+        file_format = self.session.sql(
+            f"DESCRIBE FILE FORMAT {options['FORMAT_NAME']}"
+        ).collect()
+
+        for setting in file_format:
+            if (
+                setting["property_value"] == setting["property_default"]
+                or setting["property"] in new_options
+            ):
+                continue
+
+            new_options[str(setting["property"])] = type_map.get(
+                str(setting["property_type"]), str
+            )(setting["property_value"])
+        return new_options
```

Comment:
I have some very basic questions:

1. you mentioned there are two steps -- infer schema and query, are both using the same temporary file format, or the temporary format is just for query

2. how does the code solve the parse_header issue, do we flip the value of parse_header during the query step? I don't see a special case of parse_header, if this is a generalized approach will this have side effect on other options?

3. instead of doing the fix for users, I saw doc about this parse_header, does it mean users are not supposed to set `PARSE_HEADER` when loading csv, it's their job to fix their file format?
```
This file format option is applied to the following actions only:
- Automatically detecting column definitions by using the INFER_SCHEMA function.
- Loading CSV data into separate columns by using the INFER_SCHEMA function and MATCH_BY_COLUMN_NAME copy option.
```

Reply:
1. At the moment both the infer step and the query step will create their own temporary file format if needed. This PR does not force a temporary format for inference because that step does not seem to have any incompatibilities with the external file formats.
2. This change allows the existing logic [here](https://github.com/snowflakedb/snowpark-python/blob/7f99a294aa3bb8a4f533108095083d52ba174f32/src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py#L1108-L1115) to be applied to an externally defined file format.
3. Our api only allows them to specify one file format for both inference and loading. Right now if they were to modify the file format to have parse_header be false they will not be able to load column names from the header at load time. This would is a poor experience for users.

Reply:
thanks for the explanation!